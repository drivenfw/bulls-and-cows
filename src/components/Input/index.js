import React from 'react'
import styled from 'styled-components'

import Carousel, { Value } from '../Carousel'


const StyledInput = styled.div`
  padding: 4px;
  overflow: hidden;
`

const UpButton = styled.div`
  font-size: 1.8em;
  text-align: center;
  transform: scaleX(2.2) scaleY(0.9);
  color: grey;
  cursor: pointer;
  user-select: none;
  text-shadow:
    0 2px 7px rgba(255,255,255,0.8),
    0 5px 5px rgba(0,0,0,0.4);
  transition: all 0.1s ease;

  &:active {
    transform: scaleX(2.2) scaleY(0.8) translateY(1px);
    text-shadow: 0 2px 3px rgba(0,0,0,0.3);
  }
`

const DownButton = styled.div`
  font-size: 1.8em;
  text-align: center;
  transform: rotate(180deg) scaleX(2.2) scaleY(0.8);
  color: grey;
  cursor: pointer;
  user-select: none;
  text-shadow:
    0 2px 7px rgba(255,255,255,0.8),
    0 5px 5px rgba(0,0,0,0.4);
  transition: all 0.1s ease;

  &:active {
    transform: rotate(180deg) scaleX(2.2) scaleY(0.8) translateY(1px);
    text-shadow: 0 2px 3px rgba(0,0,0,0.3);
  }
`

class Input extends React.Component {
  state = { direction: 'up', index: 0 }

  options = Array.from({ length: 9 }, (_, i) => i + 1)

  up = () => {
    this.setState(({ index }) => ({
      direction: 'down',
      index: index < this.options.length - 1 ? index + 1 : 0
    }))
  }

  down = () => {
    this.setState(({ index }) => ({
      direction: 'up',
      index: index > 0 ? index - 1 : this.options.length - 1
    })) 
  }

  get value() {
    return this.options[this.state.index]
  }

  render() {
    const { direction, index } = this.state
    const value = this.options[index]
    const valueEl = <Value key={value}>{value}</Value>

    return (
      <StyledInput>
        <UpButton onClick={this.up}>⌃</UpButton>
        <Carousel direction={direction}>
          {valueEl}
        </Carousel>
        <DownButton onClick={this.down}>⌃</DownButton>
      </StyledInput>
    )
  }
}

export default Input

