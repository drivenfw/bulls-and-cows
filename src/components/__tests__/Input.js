import React from 'react'
import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import 'jest-styled-components'

import Input from 'components/Input'


configure({ adapter: new Adapter() })

describe('<Input />', () => {
  describe('snapshot test', () => {
    it('should match the snapshot', () => {
      const tree = mount(<Input options={[0, 1, 2]} />)

      expect(toJson(tree)).toMatchSnapshot()
    })
  })
})

