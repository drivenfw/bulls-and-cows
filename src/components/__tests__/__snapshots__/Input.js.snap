// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> snapshot test should match the snapshot 1`] = `
.c3 {
  box-sizing: border-box;
  width: 35px;
  height: 60px;
  border-radius: 3px;
  font-size: 2em;
  position: relative;
  overflow: hidden;
  background: white;
  border: 1px solid;
  box-shadow: 0 0 5px 3px;
}

.c3 .scroll-down-enter {
  top: -100%;
  -webkit-transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-down-enter-active {
  top: 0;
}

.c3 .scroll-down-exit {
  top: 0;
  -webkit-transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-down-exit-active {
  top: 100%;
}

.c3 .scroll-up-enter {
  top: 100%;
  -webkit-transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-up-enter-active {
  top: 0;
}

.c3 .scroll-up-exit {
  top: 0;
  -webkit-transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: top 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-up-exit-active {
  top: -100%;
}

.c3 .scroll-left-enter {
  left: -100%;
  -webkit-transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-left-enter-active {
  left: 0;
}

.c3 .scroll-left-exit {
  left: 0;
  -webkit-transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-left-exit-active {
  left: 100%;
}

.c3 .scroll-right-enter {
  left: 100%;
  -webkit-transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-right-enter-active {
  left: 0;
}

.c3 .scroll-right-exit {
  left: 0;
  -webkit-transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
  transition: left 500ms cubic-bezier(0,0.7,0.25,1.19);
}

.c3 .scroll-right-exit-active {
  left: -100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  padding: 4px;
  overflow: hidden;
}

.c1 {
  font-size: 1.8em;
  text-align: center;
  -webkit-transform: scaleX(2.2) scaleY(0.9);
  -ms-transform: scaleX(2.2) scaleY(0.9);
  transform: scaleX(2.2) scaleY(0.9);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
  text-shadow: 0 2px 7px , 0 5px 5px;
}

.c1:active {
  -webkit-transform: scaleX(2.2) scaleY(0.8) translateY(1px);
  -ms-transform: scaleX(2.2) scaleY(0.8) translateY(1px);
  transform: scaleX(2.2) scaleY(0.8) translateY(1px);
  text-shadow: 0 2px 3px rgba(0,0,0,0.3);
}

.c5 {
  font-size: 1.8em;
  text-align: center;
  -webkit-transform: rotate(180deg) scaleX(2.2) scaleY(0.8);
  -ms-transform: rotate(180deg) scaleX(2.2) scaleY(0.8);
  transform: rotate(180deg) scaleX(2.2) scaleY(0.8);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
  text-shadow: 0 2px 7px , 0 5px 5px;
}

.c5:active {
  -webkit-transform: rotate(180deg) scaleX(2.2) scaleY(0.8) translateY(1px);
  -ms-transform: rotate(180deg) scaleX(2.2) scaleY(0.8) translateY(1px);
  transform: rotate(180deg) scaleX(2.2) scaleY(0.8) translateY(1px);
  text-shadow: 0 2px 3px rgba(0,0,0,0.3);
}

@media (min-width:375px) {
  .c0 > .c2 {
    width: 40px;
    height: 75px;
    font-size: 2.5em;
  }
}

@media (min-width:550px) {
  .c0 > .c2 {
    width: 30px;
    height: 50px;
    font-size: 2em;
  }
}

@media (min-width:850px) {
  .c0 > .c2 {
    width: 45px;
    height: 80px;
    font-size: 3em;
  }
}

@media (min-width:375px) {
  .c1 {
    font-size: 2em;
  }
}

@media (min-width:550px) {
  .c1 {
    font-size: 1.6em;
  }
}

@media (min-width:850px) {
  .c1 {
    font-size: 2.3em;
  }
}

@media (min-width:375px) {
  .c5 {
    font-size: 2em;
  }
}

@media (min-width:550px) {
  .c5 {
    font-size: 1.6em;
  }
}

@media (min-width:850px) {
  .c5 {
    font-size: 2.3em;
  }
}

<Input
  options={
    Array [
      0,
      1,
      2,
    ]
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div
        onClick={[Function]}
      >
        <div
          className="c1"
          onClick={[Function]}
        >
          ⌃
        </div>
      </styled.div>
      <Styled(Carousel)
        direction="up"
      >
        <Carousel
          className="c2 "
          direction="up"
        >
          <styled.div
            className="c2 "
          >
            <div
              className="c2 c3"
            >
              <TransitionGroup
                childFactory={[Function]}
                component="div"
              >
                <div>
                  <CSSTransition
                    classNames="scroll-up"
                    in={true}
                    key=".$.$0"
                    mountOnEnter={true}
                    onExited={[Function]}
                    timeout={500}
                    unmountOnExit={true}
                  >
                    <Transition
                      appear={false}
                      enter={true}
                      exit={true}
                      in={true}
                      mountOnEnter={true}
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      timeout={500}
                      unmountOnExit={true}
                    >
                      <styled.div
                        key="0"
                      >
                        <div
                          className="c4"
                        >
                          0
                        </div>
                      </styled.div>
                    </Transition>
                  </CSSTransition>
                </div>
              </TransitionGroup>
            </div>
          </styled.div>
        </Carousel>
      </Styled(Carousel)>
      <styled.div
        onClick={[Function]}
      >
        <div
          className="c5"
          onClick={[Function]}
        >
          ⌃
        </div>
      </styled.div>
    </div>
  </styled.div>
</Input>
`;
